### PR TITLE
[WIP] Dynamic Commands

### PR DESCRIPTION
**Pull request by gabizou**
This PR Adds the capability of dynamic commands.

 A Continuation PR of #409

### CODE REVIEW

**Comment by gabizou** [comment]
@AlphaModder while I'm getting the branch to compile, can you please explain the whole reasoning of Dynamic Commands (i.e. What are they for? What do they do? Why do we need them?)?

File: .gitignore [hunk 1]
```diff
@@ -64,6 +64,8 @@ Thumbs.db
 .idea
 .project
 .settings
+.metadata
```
Comment (Zidane) [inline .gitignore]: Doesn't belong in this PR.

**Comment by AlphaModder** [comment]
The dynamic commands idea is that it allows commands to be dynamically resolved based on their alias. Using this, the "plugin:command" notation does not need to be a duplicate of every command registered, as the code can check the prefix when the command is executed and decide which plugin to run from. This also opens up the possibility of using minecraft commands "minecraft:", and completely solves the issue of command conflicts, as a single alias could be registered by more than one plugin and each command would still be individually addressable. Finally, it also allows plugins to take control of this system, and explicitly set the plugin that will be used when a non-prefixed command is used by a certain kind of CommandSender. A way to modify this behavior through a configuration file could be in either Sponge or a sponge plugin. As for the CommandResult thing, it was just something sk89q asked me to do in this... 

tbh, with all of @boformers changes to the code, and just the general API differences, I think I should rewrite this from the ground up. It'd probably be about the same, functionality and api-structure wise, but it was my first PR, and I've learned a bit since then.

The whole thing is based on [EvilSeph's blog post](http://evilseph.com/post/75608917830/bukkit-supporting-minecraft-commands).

**Comment by boformer** [comment]
Please think about proper names for the classes and packages (not sponge...).

I also still think that ``CommandCallable.call`` should always return a ``CommandResult`` (no generics).

Maybe we can split the PR:
1. addition of ``CommandResult``
2. Command prefixes

**Comment by AlphaModder** [comment]
@boformer Yeah, if I do rewrite it I'll make CommandResult a seperate PR.

**Comment by AlphaModder** [comment]
@gabizou I assume the close is so I can open a rewritten version?

**Comment by gabizou** [comment]
Yes

**Comment by gabizou** [comment]
@boformer if you can get the CommandResult, and @AlphaModder get the dynamic command system with aliases, we can focus on the exact features. As of right now, the branch is too convoluted as to what it needs.

**Comment by boformer** [comment]
@AlphaModder is there a reason for the getAffectedBlocks/Entities/Items/Query method of CommandResult in your PR?
It seems like CommandBlocks only require the wasProcessed and getSuccessCount method.

**Comment by AlphaModder** [comment]
@boformer Those are the statistics recorded by vanilla commands.
